File: crates/tooling/chio-conformance/verdict_matrix/drivers/go/run_scenarios.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

const driverName = "go-http-sdk"
const reasonNone = "urn:chio:error:none"
const reasonScopeExceeded = "urn:chio:error:capability:scope-exceeded"
const reasonRevoked = "urn:chio:error:capability:revoked"
const reasonReplayDrift = "urn:chio:error:replay:deterministic-mismatch"
const reasonReplayTraceMissing = "urn:chio:error:replay:trace-not-found"
const reasonInputRedacted = "urn:chio:error:guard:input-redacted"
const reasonOutputRedacted = "urn:chio:error:guard:output-redacted"
const reasonGuardDenied = "urn:chio:error:guard:denied"
const reasonKernelInternal = "urn:chio:error:kernel:internal-error"
const reasonDeliveryDigestMismatch = "urn:chio:error:kernel:delivery-contract-digest-mismatch"
const reasonDeliveryUnsupportedCarrier = "urn:chio:error:kernel:delivery-contract-unsupported-carrier"

type scenario struct {
	Schema   string         `json:"schema"`
	ID       string         `json:"id"`
	Category string         `json:"category"`
	Requires []string       `json:"requires"`
	Script   scenarioScript `json:"script"`
	Expected verdictTuple   `json:"expected"`
}

type scenarioScript struct {
	Operation         string   `json:"operation"`
	Tool              string   `json:"tool"`
	InputJSON         string   `json:"input_json"`
	CapabilityScopes  []string `json:"capability_scopes"`
	RequiredScope     string   `json:"required_scope"`
	Revoked           bool     `json:"revoked"`
	ReplayNonceStatus string   `json:"replay_nonce_status"`
	RedactionAction   string   `json:"redaction_action"`
	RedactionPhase    string   `json:"redaction_phase"`
	CarrierPresent    bool     `json:"carrier_present"`
	DeliveryLane      string   `json:"delivery_lane"`
	DeliveryResult    string   `json:"delivery_result"`
}

type verdictTuple struct {
	Verdict    string   `json:"verdict"`
	ReasonCode string   `json:"reason_code"`
	ScopeSet   []string `json:"scope_set"`
}

type outcome struct {
	ScenarioID string        `json:"scenario_id"`
	Status     string        `json:"status"`
	Actual     *verdictTuple `json:"actual"`
	Expected   verdictTuple  `json:"expected"`
	Diagnostic string        `json:"diagnostic,omitempty"`
}

type report struct {
	Driver      string                  `json:"driver"`
	Total       int                     `json:"total"`
	Passed      int                     `json:"passed"`
	Failed      int                     `json:"failed"`
	Unsupported int                     `json:"unsupported"`
	Tuples      map[string]verdictTuple `json:"tuples"`
	Outcomes    []outcome               `json:"outcomes"`
}

func main() {
	root := flag.String("scenario-root", defaultScenarioRoot(), "scenario corpus root")
	flag.Parse()

	report, err := runScenarios(*root)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(report); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
	if report.Failed > 0 {
		os.Exit(1)
	}
}

func defaultScenarioRoot() string {
	root, err := repoRoot()
	if err != nil {
		return "crates/tooling/chio-conformance/verdict_matrix/scenarios"
	}
	return filepath.Join(root, "crates/tooling/chio-conformance/verdict_matrix/scenarios")
}

func repoRoot() (string, error) {
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if exists(filepath.Join(dir, "Cargo.toml")) &&
			exists(filepath.Join(dir, "crates/tooling/chio-conformance/verdict_matrix")) {
			return dir, nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", errors.New("could not find repo root")
		}
		dir = parent
	}
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func runScenarios(root string) (report, error) {
	scenarios, err := loadScenarios(root)
	if err != nil {
		return report{}, err
	}
	result := report{
		Driver:   driverName,
		Total:    len(scenarios),
		Tuples:   make(map[string]verdictTuple),
		Outcomes: make([]outcome, 0, len(scenarios)),
	}
	for _, scenario := range scenarios {
		expected := normalizeTuple(scenario.Expected)
		unsupported := unsupportedRequirement(scenario.Requires)
		if unsupported != "" {
			result.Unsupported++
			result.Outcomes = append(result.Outcomes, outcome{
				ScenarioID: scenario.ID,
				Status:     "unsupported",
				Expected:   expected,
				Diagnostic: fmt.Sprintf("unsupported requirement `%s`", unsupported),
			})
			continue
		}
		actual := evaluateScenario(scenario)
		status := "pass"
		diagnostic := ""
		if !equalTuple(actual, expected) {
			status = "fail"
			diagnostic = "go-http-sdk local semantic evaluator diverged from expected tuple"
			result.Failed++
		} else {
			result.Passed++
		}
		result.Tuples[scenario.ID] = actual
		result.Outcomes = append(result.Outcomes, outcome{
			ScenarioID: scenario.ID,
			Status:     status,
			Actual:     &actual,
			Expected:   expected,
			Diagnostic: diagnostic,
		})
	}
	return result, nil
}

func evaluateScenario(s scenario) verdictTuple {
	scopeSet := append([]string{}, s.Script.CapabilityScopes...)
	if s.Script.Operation != "tool.call" {
		return normalizeTuple(verdictTuple{
			Verdict:    "error",
			ReasonCode: reasonKernelInternal,
			ScopeSet:   scopeSet,
		})
	}

	requiredScope := s.Script.RequiredScope
	if requiredScope == "" {
		requiredScope = scopeLabelForTool(s.Script.Tool)
	}
	if s.Script.Revoked {
		return tuple("deny", reasonRevoked, scopeSet)
	}
	if !containsScope(scopeSet, requiredScope) {
		return tuple("deny", reasonScopeExceeded, scopeSet)
	}

	switch s.Category {
	case "replay":
		switch s.Script.ReplayNonceStatus {
		case "", "fresh":
			return tuple("allow", reasonNone, scopeSet)
		case "trace_missing":
			return tuple("error", reasonReplayTraceMissing, scopeSet)
		default:
			return tuple("deny", reasonReplayDrift, scopeSet)
		}
	case "redaction":
		switch s.Script.RedactionAction {
		case "deny":
			return tuple("deny", reasonGuardDenied, scopeSet)
		case "mask", "drop":
			if s.Script.RedactionPhase == "output" {
				return tuple("allow", reasonOutputRedacted, scopeSet)
			}
			return tuple("allow", reasonInputRedacted, scopeSet)
		}
	case "delivery_contract":
		// This SDK does not enforce output digests. A request carrying the
		// output_digest_sha256 constraint is denied from carrier admission
		// alone; a declared mismatch is surfaced via the digest-mismatch
		// reason, everything else via the unsupported-carrier reason.
		if s.Script.CarrierPresent {
			if s.Script.DeliveryResult == "mismatched" {
				return tuple("deny", reasonDeliveryDigestMismatch, scopeSet)
			}
			return tuple("deny", reasonDeliveryUnsupportedCarrier, scopeSet)
		}
		return tuple("allow", reasonNone, scopeSet)
	}

	return tuple("allow", reasonNone, scopeSet)
}

func tuple(verdict string, reasonCode string, scopeSet []string) verdictTuple {
	return normalizeTuple(verdictTuple{
		Verdict:    verdict,
		ReasonCode: reasonCode,
		ScopeSet:   append([]string{}, scopeSet...),
	})
}

func containsScope(scopeSet []string, required string) bool {
	for _, scope := range scopeSet {
		if scope == required {
			return true
		}
	}
	return false
}

func scopeLabelForTool(tool string) string {
	switch tool {
	case "files.read":
		return "tool:read"
	case "files.write":
		return "tool:write"
	case "system.rotate":
		return "tool:admin"
	case "metrics.query":
		return "telemetry:read"
	case "prompts.get":
		return "prompt:read"
	case "prompts.update":
		return "prompt:write"
	case "resources.read":
		return "resource:read"
	case "tools.invoke":
		return "tool:call"
	default:
		return ""
	}
}

func loadScenarios(root string) ([]scenario, error) {
	var files []string
	if err := filepath.WalkDir(root, func(path string, entry os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.IsDir() || filepath.Ext(path) != ".json" {
			return nil
		}
		files = append(files, path)
		return nil
	}); err != nil {
		return nil, err
	}
	sort.Strings(files)

	scenarios := make([]scenario, 0, len(files))
	for _, file := range files {
		bytes, err := os.ReadFile(file)
		if err != nil {
			return nil, err
		}
		var next scenario
		if err := json.Unmarshal(bytes, &next); err != nil {
			return nil, fmt.Errorf("parse %s: %w", file, err)
		}
		if next.Schema != "chio.verdict-matrix.scenario.v1" {
			return nil, fmt.Errorf("%s has unsupported scenario schema", file)
		}
		var input map[string]interface{}
		if err := json.Unmarshal([]byte(next.Script.InputJSON), &input); err != nil {
			return nil, fmt.Errorf("%s has invalid script input JSON: %w", file, err)
		}
		scenarios = append(scenarios, next)
	}
	return scenarios, nil
}

func unsupportedRequirement(requirements []string) string {
	for _, requirement := range requirements {
		switch requirement {
		case "rust-kernel", "go-http-sdk", "go-sdk", "kernel-semantics":
		default:
			return requirement
		}
	}
	return ""
}

func normalizeTuple(tuple verdictTuple) verdictTuple {
	tuple.ScopeSet = append([]string{}, tuple.ScopeSet...)
	sort.Strings(tuple.ScopeSet)
	return tuple
}

func equalTuple(left verdictTuple, right verdictTuple) bool {
	left = normalizeTuple(left)
	right = normalizeTuple(right)
	if left.Verdict != right.Verdict || left.ReasonCode != right.ReasonCode {
		return false
	}
	if len(left.ScopeSet) != len(right.ScopeSet) {
		return false
	}
	for index := range left.ScopeSet {
		if left.ScopeSet[index] != right.ScopeSet[index] {
			return false
		}
	}
	return true
}

File: crates/tooling/chio-conformance/verdict_matrix/drivers/python/run_scenarios.py
#!/usr/bin/env python3
"""Python SDK verdict-matrix driver."""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any


DRIVER_NAME = "python-sdk"
MATRIX_SERVER_ID = "verdict-matrix"

REASON_NONE = "urn:chio:error:none"
REASON_SCOPE_EXCEEDED = "urn:chio:error:capability:scope-exceeded"
REASON_REVOKED = "urn:chio:error:capability:revoked"
REASON_REPLAY_DRIFT = "urn:chio:error:replay:deterministic-mismatch"
REASON_REPLAY_TRACE_MISSING = "urn:chio:error:replay:trace-not-found"
REASON_INPUT_REDACTED = "urn:chio:error:guard:input-redacted"
REASON_OUTPUT_REDACTED = "urn:chio:error:guard:output-redacted"
REASON_GUARD_DENIED = "urn:chio:error:guard:denied"
REASON_KERNEL_INTERNAL = "urn:chio:error:kernel:internal-error"
REASON_DELIVERY_DIGEST_MISMATCH = (
    "urn:chio:error:kernel:delivery-contract-digest-mismatch"
)
REASON_DELIVERY_UNSUPPORTED_CARRIER = (
    "urn:chio:error:kernel:delivery-contract-unsupported-carrier"
)
REASON_FINDING_PURCHASE_UNSUPPORTED_ADMISSION = (
    "urn:chio:error:kernel:finding-purchase-unsupported-admission"
)
REASON_FINDING_PURCHASE_CONTEXT_INVALID = (
    "urn:chio:error:kernel:finding-purchase-context-invalid"
)
REASON_FINDING_DELIVERY_MEDIA_TYPE_MISMATCH = (
    "urn:chio:error:kernel:finding-delivery-media-type-mismatch"
)


@dataclass(frozen=True)
class VerdictTuple:
    verdict: str
    reason_code: str
    scope_set: tuple[str, ...]

    @classmethod
    def from_mapping(cls, value: dict[str, Any]) -> VerdictTuple:
        return cls(
            verdict=str(value["verdict"]),
            reason_code=str(value["reason_code"]),
            scope_set=tuple(sorted(str(scope) for scope in value["scope_set"])),
        )

    def as_json(self) -> dict[str, Any]:
        return {
            "verdict": self.verdict,
            "reason_code": self.reason_code,
            "scope_set": list(self.scope_set),
        }


def repo_root() -> Path:
    current = Path(__file__).resolve()
    for parent in current.parents:
        if (parent / "Cargo.toml").exists() and (
            parent / "crates/tooling/chio-conformance/verdict_matrix"
        ).exists():
            return parent
    raise RuntimeError(f"could not find repo root from {current}")


def install_sdk_path() -> None:
    sdk_src = repo_root() / "sdks/python/chio-sdk-python/src"
    sys.path.insert(0, str(sdk_src))


install_sdk_path()

from chio_sdk.models import (  # noqa: E402
    CapabilityToken,
    ChioScope,
    Operation,
    PromptGrant,
    ResourceGrant,
    ToolGrant,
)
from chio_sdk.testing import MockChioClient, MockVerdict  # noqa: E402


def scenario_root_default() -> Path:
    return repo_root() / "crates/tooling/chio-conformance/verdict_matrix/scenarios"


def load_scenarios(root: Path) -> list[dict[str, Any]]:
    scenarios: list[dict[str, Any]] = []
    for path in sorted(root.rglob("*.json")):
        with path.open("r", encoding="utf-8") as handle:
            scenario = json.load(handle)
        validate_scenario(scenario, path)
        scenarios.append(scenario)
    return scenarios


def validate_scenario(scenario: dict[str, Any], path: Path) -> None:
    for field in ("schema", "id", "category", "script", "expected"):
        if field not in scenario:
            raise ValueError(f"{path} missing required field {field}")
    if scenario["schema"] != "chio.verdict-matrix.scenario.v1":
        raise ValueError(f"{path} has unsupported scenario schema")
    json.loads(str(scenario["script"]["input_json"]))


def supports_requirement(requirement: str) -> bool:
    return requirement in {
        "rust-kernel",
        "python-sdk",
        "kernel-semantics",
    }


def tool_for_scope(label: str) -> str:
    mapping = {
        "tool:read": "files.read",
        "tool:write": "files.write",
        "tool:admin": "system.rotate",
        "telemetry:read": "metrics.query",
        "prompt:read": "prompts.get",
        "prompt:write": "prompts.update",
        "resource:read": "resources.read",
        "tool:call": "tools.invoke",
    }
    try:
        return mapping[label]
    except KeyError as error:
        raise ValueError(f"unsupported capability scope label `{label}`") from error


def scope_for_labels(labels: list[str]) -> ChioScope:
    grants: list[ToolGrant] = []
    resource_grants: list[ResourceGrant] = []
    prompt_grants: list[PromptGrant] = []
    for label in labels:
        tool = tool_for_scope(label)
        if label == "resource:read":
            resource_grants.append(
                ResourceGrant(uri_pattern=tool, operations=[Operation.READ])
            )
        elif label.startswith("prompt:"):
            operation = Operation.INVOKE if label == "prompt:write" else Operation.GET
            prompt_grants.append(
                PromptGrant(prompt_name=tool, operations=[operation])
            )
        else:
            grants.append(
                ToolGrant(
                    server_id=MATRIX_SERVER_ID,
                    tool_name=tool,
                    operations=[Operation.INVOKE],
                )
            )
    return ChioScope(
        grants=grants,
        resource_grants=resource_grants,
        prompt_grants=prompt_grants,
    )


def required_scope(label: str | None) -> ChioScope:
    if label is None:
        return ChioScope()
    return scope_for_labels([label])


def tuple_for(verdict: str, reason_code: str, scope_set: list[str]) -> VerdictTuple:
    return VerdictTuple(
        verdict=verdict,
        reason_code=reason_code,
        scope_set=tuple(sorted(scope_set)),
    )


def scope_label_for_tool(tool_name: str) -> str:
    mapping = {
        "files.read": "tool:read",
        "files.write": "tool:write",
        "system.rotate": "tool:admin",
        "metrics.query": "telemetry:read",
        "prompts.get": "prompt:read",
        "prompts.update": "prompt:write",
        "resources.read": "resource:read",
        "tools.invoke": "tool:call",
    }
    try:
        return mapping[tool_name]
    except KeyError as error:
        raise ValueError(f"unsupported matrix tool `{tool_name}`") from error


def unsupported_reason(scenario: dict[str, Any]) -> str | None:
    _ = scenario
    return None


def tuple_from_receipt(
    receipt: Any,
    scope_set: list[str],
) -> VerdictTuple:
    decision = receipt.decision
    if decision.verdict == "allow":
        reason_code = decision.reason or REASON_NONE
        return tuple_for("allow", reason_code, scope_set)
    if decision.verdict == "deny":
        return tuple_for(
            "deny",
            decision.reason or REASON_KERNEL_INTERNAL,
            scope_set,
        )
    return tuple_for(
        "error",
        decision.reason or REASON_KERNEL_INTERNAL,
        scope_set,
    )


async def evaluate_scenario(scenario: dict[str, Any]) -> VerdictTuple:
    script = scenario["script"]
    scope_set = list(script.get("capability_scopes", []))
    if script.get("operation") != "tool.call":
        return tuple_for("error", REASON_KERNEL_INTERNAL, scope_set)

    required_label = str(
        script.get("required_scope") or scope_label_for_tool(str(script["tool"]))
    )
    if script.get("revoked", False):
        return tuple_for("deny", REASON_REVOKED, scope_set)
    if required_label not in scope_set:
        return tuple_for("deny", REASON_SCOPE_EXCEEDED, scope_set)

    category = str(scenario["category"])
    if category == "replay":
        nonce_status = str(script.get("replay_nonce_status", "fresh"))
        if nonce_status == "fresh":
            return tuple_for("allow", REASON_NONE, scope_set)
        if nonce_status == "trace_missing":
            return tuple_for("error", REASON_REPLAY_TRACE_MISSING, scope_set)
        return tuple_for("deny", REASON_REPLAY_DRIFT, scope_set)

    if category == "redaction":
        action = str(script.get("redaction_action", "none"))
        phase = str(script.get("redaction_phase", "input"))
        if action == "deny":
            return tuple_for("deny", REASON_GUARD_DENIED, scope_set)
        if action in {"mask", "drop"}:
            reason = (
                REASON_OUTPUT_REDACTED
                if phase == "output"
                else REASON_INPUT_REDACTED
            )
            return tuple_for("allow", reason, scope_set)

    if category == "delivery_contract":
        # This SDK does not enforce output digests. A request carrying the
        # output_digest_sha256 constraint is denied from carrier admission
        # alone; the scenario declares the delivery ground truth so a
        # mismatch is surfaced distinctly from an unenforceable carrier.
        if bool(script.get("carrier_present", False)):
            if str(script.get("delivery_result", "none")) == "mismatched":
                return tuple_for(
                    "deny", REASON_DELIVERY_DIGEST_MISMATCH, scope_set
                )
            return tuple_for(
                "deny", REASON_DELIVERY_UNSUPPORTED_CARRIER, scope_set
            )
        return tuple_for("allow", REASON_NONE, scope_set)

    if category == "finding_purchase":
        # This SDK holds none of the authorities a marked reveal needs: it
        # cannot verify a signed purchase context, so a grant carrying the
        # require_finding_purchase marker can only fail closed. The
        # scenario declares the marker and context ground truth so a
        # rejected purchase stays distinct from an unenforceable marker.
        if not bool(script.get("marker_present", False)):
            return tuple_for("allow", REASON_NONE, scope_set)
        if (
            str(script.get("marker_state", "well_formed")) != "well_formed"
            or not bool(script.get("purchase_context_present", False))
            or not bool(script.get("token_matches_offer", False))
            or not bool(script.get("argument_matches_marker", False))
        ):
            return tuple_for(
                "deny", REASON_FINDING_PURCHASE_CONTEXT_INVALID, scope_set
            )
        if str(script.get("media_type_result", "none")) == "mismatched":
            return tuple_for(
                "deny", REASON_FINDING_DELIVERY_MEDIA_TYPE_MISMATCH, scope_set
            )
        return tuple_for(
            "deny", REASON_FINDING_PURCHASE_UNSUPPORTED_ADMISSION, scope_set
        )

    return tuple_for("allow", REASON_NONE, scope_set)


async def run_scenarios(root: Path) -> dict[str, Any]:
    outcomes = []
    tuples: dict[str, Any] = {}
    for scenario in load_scenarios(root):
        expected = VerdictTuple.from_mapping(scenario["expected"])
        unsupported = [
            requirement
            for requirement in scenario.get("requires", [])
            if not supports_requirement(str(requirement))
        ]
        scenario_unsupported = unsupported_reason(scenario)
        if unsupported:
            actual = None
            status = "unsupported"
            diagnostic = f"unsupported requirement `{unsupported[0]}`"
        elif scenario_unsupported is not None:
            actual = None
            status = "unsupported"
            diagnostic = scenario_unsupported
        else:
            actual_tuple = await evaluate_scenario(scenario)
            actual = actual_tuple.as_json()
            status = "pass" if actual_tuple == expected else "fail"
            diagnostic = None
            tuples[scenario["id"]] = actual
        outcomes.append(
            {
                "scenario_id": scenario["id"],
                "status": status,
                "actual": actual,
                "expected": expected.as_json(),
                "diagnostic": diagnostic,
            }
        )
    failed = sum(1 for outcome in outcomes if outcome["status"] == "fail")
    unsupported_count = sum(
        1 for outcome in outcomes if outcome["status"] == "unsupported"
    )
    return {
        "driver": DRIVER_NAME,
        "total": len(outcomes),
        "passed": len(outcomes) - failed - unsupported_count,
        "failed": failed,
        "unsupported": unsupported_count,
        "tuples": tuples,
        "outcomes": outcomes,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--scenario-root",
        type=Path,
        default=scenario_root_default(),
    )
    args = parser.parse_args()
    report = asyncio.run(run_scenarios(args.scenario_root))
    print(json.dumps(report, sort_keys=True, indent=2))
    return 1 if report["failed"] else 0


if __name__ == "__main__":
    raise SystemExit(main())
